python/mxnet/executor_manager.py

```diff
@@ -0,0 +1,368 @@
+    def __init__(self, sym, arg_names, param_names, ctx, slices, train_data, shared_group=None):
+        # make sure the architecture is valid
+        _check_arguments(sym)
+
+        if shared_group is None:
+            self.shared_data_arrays = [{} for _ in ctx]
+        else:
+            self.shared_data_arrays = shared_group.shared_data_arrays
+
+        self.data_names = [x[0] for x in train_data.provide_data]
+        self.label_names = [x[0] for x in train_data.provide_label]
+        self.aux_names = sym.list_auxiliary_states()
+        self.param_idx = [i for i in range(len(arg_names)) if arg_names[i] in param_names]
+        self.param_names = [arg_names[i] for i in self.param_idx]
+
+        self.train_execs = []
+        for i in range(len(ctx)):
+            data_shapes = {k: tuple([slices[i].stop-slices[i].start] + list(v[1:]))
+                           for k, v in train_data.provide_data + train_data.provide_label}
+            shared_exec = None if shared_group is None else shared_group.train_execs[i]
+            train_exec = _bind_exec(sym, ctx[i], data_shapes, self.param_names,
+                                    need_grad=True, base_exec=shared_exec,
+                                    shared_data_arrays=self.shared_data_arrays[i])
+            self.train_execs.append(train_exec)
+
+        # data structure
+        self.data_arrays = [[(slices[i], e.arg_dict[name]) for i, e in enumerate(self.train_execs)]
+                            for name in self.data_names]
+        self.label_arrays = [[(slices[i], e.arg_dict[name]) for i, e in enumerate(self.train_execs)]
+                             for name in self.label_names]
+
+        self.param_arrays = [[e.arg_arrays[i] for e in self.train_execs]
+                             for i in self.param_idx]
+        self.grad_arrays = [[e.grad_arrays[i] for e in self.train_execs]
+                            for i in self.param_idx]
+
+        self.aux_arrays = [[e.aux_arrays[i] for e in self.train_execs]
+                           for i in range(len(self.aux_names))]
+
+        self.slices = slices
+
+    def load_data_batch(self, data_batch):
+        """ load data and labels into arrays """
+        _load_data(data_batch, self.data_arrays)
+        _load_label(data_batch, self.label_arrays)
+
+    def forward(self, is_train=False):
+        """ Perform a forward pass on each executor """
+        for texec in self.train_execs:
+            texec.forward(is_train=is_train)
+
+    def backward(self):
+        """ Perform a backward pass on each executor """
+        for texec in self.train_execs:
+            texec.backward()
+
+    def update_metric(self, metric, labels):
+        """ Update evaluation metric with label and current outputs """
+        for texec, islice in zip(self.train_execs, self.slices):
+            labels_slice = [label[islice] for label in labels]
+            metric.update(labels_slice, texec.outputs)
+
+class DataParallelExecutorManager(object):
```

Comment:
I'm open to better naming. But I think the inheritance makes things confusing, as they are complete different concepts. Current the manager keeps reference to multiple groups. The delegate API call just delegate to the *current* group.